File: 2074.reverse-nodes-in-even-length-groups.java
#
# @lc app=leetcode id=2074 lang=java
#
# [2074] Reverse Nodes in Even Length Groups
#

# @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode reverseEvenLengthGroups(ListNode head) {
        ListNode dummy = new ListNode(0, head);
        ListNode prevTail = dummy;
        int groupSize = 1;
        while (prevTail.next != null) {
            ListNode groupStart = prevTail.next;
            ListNode groupTail = null;
            ListNode temp = groupStart;
            int length = 0;
            while (temp != null && length < groupSize) {
                length++;
                groupTail = temp;
                temp = temp.next;
            }
            ListNode nextGroupHead = temp;
            if (length % 2 == 0) {
                // Iterative reverse
                ListNode revPrev = nextGroupHead;
                ListNode revCurr = groupStart;
                while (revCurr != nextGroupHead) {
                    ListNode revNextTemp = revCurr.next;
                    revCurr.next = revPrev;
                    revPrev = revCurr;
                    revCurr = revNextTemp;
                }
                prevTail.next = revPrev;
                prevTail = groupStart;
            } else {
                prevTail = groupTail;
            }
            groupSize++;
        }
        return dummy.next;
    }
}
# @lc code=end

File: 2074.reverse-nodes-in-even-length-groups.go
#
# @lc app=leetcode id=2074 lang=golang
#
# [2074] Reverse Nodes in Even Length Groups
#

# @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseEvenLengthGroups(head *ListNode) *ListNode {
	dummy := &ListNode{0, head}
	prevEnd := dummy
	groupSize := 1
	for prevEnd.Next != nil {
		groupStart := prevEnd.Next
		cur := groupStart
		groupLen := 0
		for groupLen < groupSize && cur != nil {
			cur = cur.Next
			groupLen++
		}
		if groupLen%2 == 0 {
			newHead := (*ListNode)(nil)
			curNode := groupStart
			for i := 0; i < groupLen; i++ {
				next := curNode.Next
				curNode.Next = newHead
				newHead = curNode
				curNode = next
			}
			prevEnd.Next = newHead
			groupStart.Next = cur
			prevEnd = groupStart
		} else {
			groupEnd := groupStart
			for i := 1; i < groupLen; i++ {
				groupEnd = groupEnd.Next
			}
			prevEnd = groupEnd
		}
		groupSize++
	}
	return dummy.Next
}

# @lc code=end